libs/community/langchain_community/storage/astradb.py

```diff
@@ -0,0 +1,115 @@
+    def __init__(
+        self,
+        collection_name: str,
+        token: Optional[str] = None,
+        api_endpoint: Optional[str] = None,
+        astra_db_client: Optional[Any] = None,  # 'astrapy.db.AstraDB' if passed
+        namespace: Optional[str] = None,
+    ) -> None:
+        try:
+            from astrapy.db import AstraDB, AstraDBCollection
+        except (ImportError, ModuleNotFoundError):
+            raise ImportError(
+                "Could not import a recent astrapy python package. "
+                "Please install it with `pip install --upgrade astrapy`."
+            )
+
+        # Conflicting-arg checks:
+        if astra_db_client is not None:
+            if token is not None or api_endpoint is not None:
+                raise ValueError(
+                    "You cannot pass 'astra_db_client' to AstraDB if passing "
+                    "'token' and 'api_endpoint'."
+                )
+
+        astra_db = astra_db_client or AstraDB(
+            token=token,
+            api_endpoint=api_endpoint,
+            namespace=namespace,
+        )
+        self.collection = AstraDBCollection(collection_name, astra_db=astra_db)
+
+    @abstractmethod
+    def decode_value(self, value: Any) -> Optional[V]:
+        """Decodes value from Astra DB"""
+
+    @abstractmethod
+    def encode_value(self, value: Optional[V]) -> Any:
+        """Encodes value for Astra DB"""
+
+    def mget(self, keys: Sequence[str]) -> List[Optional[V]]:
+        """Get the values associated with the given keys."""
+        docs_dict = {}
+        for doc in self.collection.paginated_find(filter={"_id": {"$in": list(keys)}}):
+            docs_dict[doc["_id"]] = doc.get("value")
+        return [self.decode_value(docs_dict.get(key)) for key in keys]
+
+    def mset(self, key_value_pairs: Sequence[Tuple[str, V]]) -> None:
+        """Set the given key-value pairs."""
+        docs = []
+        for k, v in key_value_pairs:
+            docs.append({"_id": k, "value": self.encode_value(v)})
+            if len(docs) == 20:
+                self.collection.insert_many(docs)
+                docs.clear()
+        if len(docs) > 0:
+            self.collection.insert_many(docs)
+
+    def mdelete(self, keys: Sequence[str]) -> None:
+        """Delete the given keys."""
+        self.collection.delete_many(filter={"_id": {"$in": list(keys)}})
+
+    def yield_keys(self, *, prefix: Optional[str] = None) -> Iterator[str]:
+        """Yield keys in the store."""
+        docs = self.collection.paginated_find()
+        for doc in docs:
+            key = doc["_id"]
+            if not prefix or key.startswith(prefix):
+                yield key
+
+
+class AstraDBStore(AstraDBBaseStore[Any]):
+    """BaseStore implementation using DataStax AstraDB as the underlying store.
+    The value type can be any type serializable by json.dumps.
+    Can be used to store embeddings with the CacheBackedEmbeddings.
```

Comment:
Can we add an example of the output ? is it a dict with column names as key ? an array ?

Reply:
done